development: strict types could be returned from handlers

onokonem opened the pull request:
to make code more safe

casualjim commented:
Do you have an example of what the end result looks like? Maybe in the examples folder?

onokonem commented:
of course I do :)


[configure_issue_tracker.go](https://github.com/go-swagger/go-swagger/files/2975515/configure_issue_tracker.go.txt)

this is a file generated from `examples/task-tracker`. there are  some more changes but they all are related to the changes in this file.

The main idea is to force the handler to return one of the replays predefined for this particular method.

onokonem commented:
and yes, I can create a strict (and build api) examples based on the existing ones. just tell me which one suits best

casualjim commented:
Either the todo list api or the issue tracker example would be great

onokonem commented:
done

onokonem commented:
BTW, I need help adding tests for these options.

casualjim commented:
so essentially the change is it returns the more precise responder types instead of the interface?
And the builder changes because of that, including the not implemented error?

onokonem commented:
exactly

casualjim commented:
for tests we typically take a spec, generate code and verify that the files have the expected content through regex matchers etc.

casualjim commented:
I suppose this is not ready for merge yet is that correct?

onokonem commented:
from my point of view it is ready to be merged

it is in use in prod in some place at the moment

the parts are missing:

- tests to satisfy the coverage checker
- tests to actual test the functionality

unfortunately it looks like I'll not complete the tests

casualjim commented:
Unfortunately we need tests to accept this contribution, so that it doesn't inadvertently breaks. The last year we spent increasing our test coverage and stability.

onokonem commented:
> Unfortunately we need tests to accept this contribution, so that it doesn't inadvertently breaks. The last year we spent increasing our test coverage and stability.

do you have your test system documented somewhere?

I might be able to assign somebody less busy/lazy to this task

casualjim commented:
I don't think we have it documented somewhere but it's pretty straight forward. You make a spec or take one of the pre-existing ones, and then call render. We have tons of tests that all look like this:

https://github.com/go-swagger/go-swagger/blob/master/generator/discriminators_test.go#L38-L65

Or more in line with this changeset:
https://github.com/go-swagger/go-swagger/blob/master/generator/server_test.go#L347-L379

fredbi commented:
@onokonem I've had a look at this work and was surprised about the way you are duplicating a lot of templates.
Is this change so hard on response, etc templates so as to require an entire copy of this source?

Couldn't the same be achieved with proper {{ if .Stricts }} ... {{ else }} ... {{ end }} with a single template?

onokonem commented:
> Couldn't the same be achieved with proper {{ if .Stricts }} ... {{ else }} ... {{ end }} with a single template?

@fredbi , of course it could

But these templates are overcomplicated already.

fredbi commented:
> But these templates are overcomplicated already.

So you think that duplicating the source would make the overall solution simpler to maintain and safer to use? I don't think anyone is going to apply their changes everywhere and maintain both sets of templates.

Another way to "fork out" templates is to add "contrib" templates, which live their own lives next to the main templates set.

onokonem commented:
> Another way to "fork out" templates is to add "contrib" templates, which live their own lives next to the main templates set

I've tried that first, but found I need one more parameter passed to templates.

So I've added them

For my own task I've added the templates we are using to the project.

And yes, I'm ready to discuss how this has to be refactored to be added to the upstream, as it is a little pain to maintain the fork.

fredbi commented on generator/shared.go:
```diff
@@ -580,6 +588,9 @@ type GenOpts struct {
 	ExistingModels         string
 	Copyright              string
+
+	Strict          bool
```
GenOpts are available from the template, so there is no need to duplicate templates

fredbi commented on generator/shared.go:
```diff
@@ -1257,3 +1268,10 @@ func sharedValidationsFromSchema(v spec.Schema, isRequired bool) (sh sharedValid
 	return
 }
+
+func stringTernary(cond bool, ifTrue string, ifFalse string) string { // nolint: unparam
```
the // nolint: hint is in my opinion unnecessary. You could just declare this has: (cond bool, ifTrue, ifFalse string) string

onokonem commented:
great, will do on weekend

onokonem commented:
replaced by https://github.com/go-swagger/go-swagger/pull/2312
